lib/confex/resolver.ex

```diff
@@ -0,0 +1,134 @@
+defmodule Confex.Resolver do
+  @moduledoc """
+  This module provides API to recursively resolve Confex configuration types in a `Map` or `Keyword` via
+  one of `Confex.Adpater`'s.
+  """
+  alias Confex.Adapter
+  alias Confex.Type
+
+  @types [:string, :integer, :float, :boolean, :atom, :module, :list]
+
+  # If you are distributing new adapter for Confex,
+  # please contribute additional value to this list.
+  @adapters Application.fetch_env!(:confex, :adapters)
```

Comment:
This was a hard decision for me. This happens because:

I need to recursively walk on keyword lists. Their elements represented internally as `{:atom, _value}`, which conflicts with `{:system, "ENV_NAME"}` notation. To solve this dilema I've decided to keep hardcoded list of adapters and added guard to match on first tuple element,.

Values itself moved to Application settings, so I will be possible to update this list before compiling Confex dependency.

Probably this wasn't the best solution, but I didn't find anything better.